### ISSUE TITLE
version 2 breaks quite a lot

### ISSUE THREAD

**Issue by kklepper** [docker/compose#3127]
example: 

```
git clone https://github.com/tutumcloud/hello-world.git
docker build -t hello-world .
```

add docker-compose.yml

```
version: '2'
services:
  web:
    image: hello-world
    ports:
      - "80:80"
    links:
      - mem
  mem:
    image: memcached 
```

run:

```
$ docker-compose up -d && sudo docker-compose ps
Creating network "helloworld2_default" with the default driver
Creating helloworld2_mem_1
Creating helloworld2_web_1
      Name                     Command               State         Ports
-------------------------------------------------------------------------------
helloworld2_mem_1   /entrypoint.sh memcached         Up      11211/tcp
helloworld2_web_1   /bin/sh -c php-fpm -d vari ...   Up      0.0.0.0:80->80/tcp
```

Now point your browser to the right address. You'd expect to see links here - but you don't. 

```
Hello world!
My hostname is 79c7b04ff3bd         
```

Comment the first 2 lines out, then you'll see the links you expect.

```
Hello world!
My hostname is caafd3f11f11
Links found
HELLOWORLD2_MEM_1 listening in 11211 available at tcp://172.17.0.2:11211
MEM listening in 11211 available at tcp://172.17.0.2:11211
MEM_1 listening in 11211 available at tcp://172.17.0.2:11211
```

As a consequence, if you have a fairly complicated setup with lots of links, they can't connect as advertised with version 2. 

Am I missing something here?

docker

```
$ docker info
Containers: 6
 Running: 5
 Paused: 0
 Stopped: 1
Images: 53
Server Version: 1.10.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 111
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 3.13.0-79-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 992.5 MiB
Name: ip-172-31-30-16
ID: YCVV:YNAX:447K:W5Y7:UMCK:QIJB:3QMI:N47Z:Q2UH:IMFW:HQXA:GWIO
WARNING: No swap limit support
```

compose

```
$ docker-compose version
docker-compose version 1.6.2, build 4d72027
docker-py version: 1.7.2
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1e 11 Feb 2013
```

**Comment by dnephin** [comment]
Yes, links are no longer provided by `/etc/hosts`, they are now provided by an embedded DNS server available in Docker 1.10.x. This gives you network-scoped aliases as well, so you don't need explicit links between services for them to be discoverable.

If you want the old behaviour you can use `network_mode: bridge` in every service, but you won't get the network scoped links.

**Comment by kklepper** [comment]
Thank you very much.

**Comment by kklepper** [comment]
May I ask you another question, please. I really struggle translating my docker-compose setup to the new version. According to the blog entry, it should rather work out of the box.

To make my example more interesting, I replaced memcached with a MariaDB database. Using version 1, I had no problem connecting to this database and working with it. Using version 2, I have no idea what to do. I tried a couple of things, to no avail.

    version: '2'
        services:
          web:
            image: hello-world
            container_name: web
            volumes:
              - ./www/:/www/
            ports:
              - "80:80"
            links:
              - db
          db:
            image: mariadb
            container_name: db
            ports:
              - "3306"
            volumes_from:
              - data
            environment:
              - MYSQL_ROOT_PASSWORD=*********
          data:
            image: gliderlabs/alpine
            container_name: dat
            command:
              - 'true'
            volumes:
              - ~/wp/data:/var/lib/mysql
        
This is the result:

    Warning: mysqli_connect(): (HY000/1130): Host '172.18.0.3' is not allowed to connect to this MariaDB server in /www/index.php on line 65

Actually the database container runs on 172.18.0.2. Commenting out the first 2 lines, everything is fine:

    Select returned 15 rows
    DB listening in 3306 available at tcp://172.17.0.2:3306
    
What is it that I am doing wrong? What should I do instead?

**Comment by dnephin** [comment]
It could be that the mariadb image only allows users from the `172.17.x.x` subnet? Maybe the user permissions need to be updated to allow them from all subnets?

**Comment by kklepper** [comment]
@dnephin 
bind-address            = 0.0.0.0

**Comment by kklepper** [comment]
My question can be simplified. The following scenario will not work:

    version: '2'
    services:
    
      web:
        image:  dockercloud/hello-world
        expose:
          - 5000
      lb:
        image: dockercloud/haproxy
        links:
          - web
        ports:
          - "80:80"
          - "1936:1936"                              # for stats
        expose:
          - 80
        environment:
          - BACKEND_PORT=5000
          - BALANCE=roundrobin
          - STATS_AUTH=admin:admin

The first steps look fine:
    
    ubuntu@ip-172-31-30-16:~/docker/dockercloud-hello-world$ sudo docker-compose up -d && sudo docker-compose ps
    Creating network "dockercloudhelloworld_default" with the default driver
    Creating dockercloudhelloworld_web_1
    Creating dockercloudhelloworld_lb_1
               Name                          Command               State                          Ports
    --------------------------------------------------------------------------------------------------------------------------
    dockercloudhelloworld_lb_1    tini -- dockercloud-haproxy      Up      0.0.0.0:1936->1936/tcp, 443/tcp, 0.0.0.0:80->80/tcp
    dockercloudhelloworld_web_1   /bin/sh -c php-fpm -d vari ...   Up      5000/tcp, 80/tcp
    
I am on AWS EC2 Ubuntu t2.micro. My browser says no connection. The same holds true locally:

    ubuntu@ip-172-31-30-16:~/docker/dockercloud-hello-world$ curl localhost
    curl: (52) Empty reply from server
    ubuntu@ip-172-31-30-16:~/docker/dockercloud-hello-world$ sudo docker inspect -f '{{ .NetworkSettings.Networks.dockercloudhelloworld_default.IPAddress }}' dockercloudhelloworld_lb_1
    172.20.0.3
    ubuntu@ip-172-31-30-16:~/docker/dockercloud-hello-world$ curl 172.20.0.3
    curl: (7) Failed to connect to 172.20.0.3 port 80: Connection refused

Now if you omit

    version: '2'
    services:

everything works as expected. What is this, please? Am I missing something?

(PS: dockercloud/hello-world misses the image `logo.png` shown in tutum/hello-world)

**Comment by kklepper** [comment]
Another example, taken from https://github.com/SirIle/miniboxes/tree/master/minihaproxy, ported to docker-compose version 1/2

Define some constants

    export DOCKER_MACHINE_NAME=$(echo $SSH_CONNECTION | awk {'print $3'}) 
    export DOCKER_BASE_IP=$(ifconfig docker0 | grep 'inet addr'| awk -F':' '{print $2}' | awk '{print $1}') 


docker-compose.yml:

    version: '2'
    services:
      
      consul:
        command: -server -bootstrap-expect 1
        image: progrium/consul:latest
        container_name: consul
        ports:
          - "8300:8300"
          - "8301:8301"
          - "8500:53/udp"
          - "8302:8302"
          - "8400:8400"
          - "8500:8500"
          - "8600:8600"
          - "$DOCKER_BASE_IP:53:53/udp"                 # does not need to be docker0 inet addr:172.17.0.1
      
      registrator:
        command: -ip="$DOCKER_MACHINE_NAME" -internal consul://consul:8500  # must be private IP
        image: gliderlabs/registrator:latest
        container_name: reg
        volumes:
          - "/var/run/docker.sock:/tmp/docker.sock"
        links:
          - consul
      
      hello:
        image: sirile/scala-boot-test
        ports:
          - "80"
        environment:
          SERVICE_NAME: hello/v1
          SERVICE_TAGS: rest
      
      rest:
        image: sirile/haproxy
        dns: $DOCKER_BASE_IP
        container_name: lb
        links:
          - consul
        ports:
          - "80:80"
          - "1936:1936"
        environment:
          SERVICE_NAME: rest

Start the stack:

    $ docker-compose up -d && docker-compose ps     

Now check for haproxy.cfg:

    $  docker run --dns 172.17.0.2 --rm sirile/haproxy -consul=consul.service.consul:8500 -dry -once
    2016/04/14 18:42:22 [ERR] (view) "storeKeyPrefix(service/haproxy/timeouts)" store key prefix: error fetching: Get http://consul.service.consul:8500/v1/kv/service/haproxy/timeouts?recurse=&stale=&wait=60000ms: dial tcp: lookup consul.service.consul on 172.17.0.2:53: read udp 172.17.0.2:57997->172.17.0.2:53: read: connection refused
    2016/04/14 18:42:22 [ERR] (runner) watcher reported error: store key prefix: error fetching: Get http://consul.service.consul:8500/v1/kv/service/haproxy/timeouts?recurse=&stale=&wait=60000ms: dial tcp: lookup consul.service.consul on 172.17.0.2:53: read udp 172.17.0.2:57997->172.17.0.2:53: read: connection refused
    Consul Template returned errors:
    store key prefix: error fetching: Get http://consul.service.consul:8500/v1/kv/service/haproxy/timeouts?recurse=&stale=&wait=60000ms: dial tcp: lookup consul.service.consul on 172.17.0.2:53: read udp 172.17.0.2:57997->172.17.0.2:53: read: connection refusedubuntu@ip-172-31-30-16:~/kklepper/voxx/consul$

Comment the first two lines in docker-compose.yml to return back to version 1, restart the stack and test again:

	$ docker-compose down && docker-compose up -d && docker-compose ps 

This is what it should look like:	

    $ docker run --dns 172.17.0.2 --rm sirile/haproxy -consul=consul.service.consul:8500 -dry -once
    2016/04/14 18:55:12 [WARN] ("key(service/haproxy/maxconn)") Consul returned no data (does the path exist?)
    > /etc/haproxy/haproxy.cfg
    global
        log 127.0.0.1   local0
        log 127.0.0.1   local1 notice
        debug
        stats timeout 30s
        maxconn 4096
    
    defaults
        log global
        option httplog
        option dontlognull
        mode http
        timeout connect 5000
        timeout client  50000
        timeout server  50000
    
    frontend http-in
        bind *:80
        acl app8 path_beg -i /hello/v1
    
        use_backend srvs_app8 if app8
    
    
    backend srvs_app8
        mode http
        balance roundrobin
        option forwardfor
        option httpchk HEAD / HTTP/1.1\r\nHost:localhost
        reqrep ^([^\ ]*\ /)hello/v1[/]?(.*)     \1\2
        server host0 172.17.0.4:80 check
    
    listen stats
        bind *:1936
        stats enable
        stats uri /
        stats hide-version
        stats auth someuser:password

In case of version 2, the log file of the load balancer looks normal at startup:

    $ docker logs lb
    2016/04/14 18:56:29 [ERR] (view) "services" catalog services: error fetching: Get http://consul.service.consul:8500/v1/catalog/services?stale=&wait=60000ms: dial tcp: lookup consul.service.consul on 127.0.0.11:53: server misbehaving
    2016/04/14 18:56:29 [ERR] (runner) watcher reported error: catalog services: error fetching: Get http://consul.service.consul:8500/v1/catalog/services?stale=&wait=60000ms: dial tcp: lookup consul.service.consul on 127.0.0.11:53: server misbehaving
    2016/04/14 18:56:29 [ERR] (view) "key(service/haproxy/maxconn)" store key: error fetching: Get http://consul.service.consul:8500/v1/kv/service/haproxy/maxconn?stale=&wait=60000ms: dial tcp: lookup consul.service.consul on 127.0.0.11:53: server misbehaving
    2016/04/14 18:56:29 [ERR] (runner) watcher reported error: store key: error fetching: Get http://consul.service.consul:8500/v1/kv/service/haproxy/maxconn?stale=&wait=60000ms: dial tcp: lookup consul.service.consul on 127.0.0.11:53: server misbehaving
    2016/04/14 18:56:29 [ERR] (view) "storeKeyPrefix(service/haproxy/timeouts)" store key prefix: error fetching: Get http://consul.service.consul:8500/v1/kv/service/haproxy/timeouts?recurse=&stale=&wait=60000ms: dial tcp: lookup consul.service.consul on 127.0.0.11:53: server misbehaving
    2016/04/14 18:56:29 [ERR] (runner) watcher reported error: store key prefix: error fetching: Get http://consul.service.consul:8500/v1/kv/service/haproxy/timeouts?recurse=&stale=&wait=60000ms: dial tcp: lookup consul.service.consul on 127.0.0.11:53: server misbehaving
    2016/04/14 18:56:34 [WARN] ("key(service/haproxy/maxconn)") Consul returned no data (does the path exist?)
    Using epoll() as the polling mechanism.
    [WARNING] 104/185634 (13) : config : frontend 'GLOBAL' has no 'bind' directive. Please declare it as a backend if this was intended.
    Available polling systems :
          epoll : pref=300,  test result OK
           poll : pref=200,  test result OK
         select : pref=150,  test result FAILED
    Total: 3 (2 usable), will use epoll.
    [WARNING] 104/185634 (13) : Server srvs_app8/host0 is DOWN, reason: Layer4 connection problem, info: "Connection refused", check duration: 0ms. 0 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
    [ALERT] 104/185634 (13) : backend 'srvs_app8' has no server available!

But `curl localhost/hello/v1` shows that we won't get data:

    $ curl localhost/hello/v1
    <html><body><h1>503 Service Unavailable</h1>
    No server is available to handle this request.
    </body></html>

Now the log file gives us an error -- we don't have a server and we don't get one:

    [ALERT] 104/185634 (13) : backend 'srvs_app8' has no server available!
    00000000:http-in.accept(0004)=0007 from [172.18.0.1:59243]
    00000000:http-in.clireq[0007:ffffffff]: GET /hello/v1 HTTP/1.1
    00000000:http-in.clihdr[0007:ffffffff]: User-Agent: curl/7.35.0
    00000000:http-in.clihdr[0007:ffffffff]: Host: localhost
    00000000:http-in.clihdr[0007:ffffffff]: Accept: */*
    00000000:srvs_app8.clicls[0007:ffffffff]
    00000000:srvs_app8.closed[0007:ffffffff]
    2016/04/14 18:57:37 [WARN] ("key(service/haproxy/maxconn)") Consul returned no data (does the path exist?)
    2016/04/14 18:58:38 [WARN] ("key(service/haproxy/maxconn)") Consul returned no data (does the path exist?)
    2016/04/14 18:59:39 [WARN] ("key(service/haproxy/maxconn)") Consul returned no data (does the path exist?)
    00000001:http-in.accept(0004)=0007 from [172.18.0.1:59325]
    00000001:http-in.clireq[0007:ffffffff]: GET /hello/v1 HTTP/1.1
    00000001:http-in.clihdr[0007:ffffffff]: User-Agent: curl/7.35.0
    00000001:http-in.clihdr[0007:ffffffff]: Host: localhost
    00000001:http-in.clihdr[0007:ffffffff]: Accept: */*
    00000001:srvs_app8.clicls[0007:ffffffff]
    00000001:srvs_app8.closed[0007:ffffffff]

It should have been instead:

    [ALERT] 104/190241 (14) : backend 'srvs_app8' has no server available!
    [WARNING] 104/190250 (14) : Server srvs_app8/host0 is UP, reason: Layer7 check passed, code: 200, info: "OK", check duration: 5ms. 1 active and 0 backup servers online. 0 sessions requeued, 0 total in queue.
    00000000:http-in.accept(0004)=0007 from [172.17.0.1:39665]
    00000000:http-in.clireq[0007:ffffffff]: GET /hello/v1 HTTP/1.1
    00000000:http-in.clihdr[0007:ffffffff]: User-Agent: curl/7.35.0
    00000000:http-in.clihdr[0007:ffffffff]: Host: localhost
    00000000:http-in.clihdr[0007:ffffffff]: Accept: */*
    00000000:srvs_app8.srvrep[0007:0008]: HTTP/1.1 200 OK
    00000000:srvs_app8.srvhdr[0007:0008]: Date: Thu, 14 Apr 2016 19:03:08 GMT
    00000000:srvs_app8.srvhdr[0007:0008]: X-Application-Context: application:80
    00000000:srvs_app8.srvhdr[0007:0008]: Content-Type: application/json;charset=UTF-8
    00000000:srvs_app8.srvhdr[0007:0008]: Transfer-Encoding: chunked
    00000000:srvs_app8.srvhdr[0007:0008]: Server: Jetty(9.2.9.v20150224)
    00000001:http-in.clicls[0007:0008]
    00000001:http-in.closed[0007:0008]

Line 2 shows that the appropriate server has been fired up to deliver data:

    $ curl localhost/hello/v1
    {"hostname":"4ee0cfe39fa2","time":"2016-04-14T19:09:01Z","language":"Scala"}		

So again, either something is utterly wrong with version 2 or I must miss something which should be obvious.

**Comment by vguna** [comment]
I encounter the same problem. For me it seems to have something todo with the reqrep line. If I remove it, it seems not to fail. Did you ever solve this?
